File: shttp.Tests/FullRequestTests/UserCacheTests.cs
using Moq;
using NUnit.Framework;
using shttp.Tests.TestUtils;
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;

namespace shttp.Tests
{
    public class UserCacheTests
    {
        [Test]
        public async Task ClientRequest_WithUserHeaderMaxAgeAndPublic_AddsToSharedCache()
        {
            // arrange
            var state = new TestState();
            var expectedResponse = state.AddHttpRequest(1);
            expectedResponse.Headers.CacheControl = new CacheControlHeaderValue
            {
                MaxAge = TimeSpan.FromDays(1),
                Private = false
            };

            // act
            var response = await state.ExecuteRequest(user: "my user");

            // assert
            Predicate<Tuple<string, HttpResponseMessage, DateTime>> assert = AssertResult;
            state.Dependencies
                .Verify(x => x.Cache.Put(Match.Create(assert)), Times.Once);

            bool AssertResult(Tuple<string, HttpResponseMessage, DateTime> input)
            {
                Assert.AreEqual("$:$:http://www.com/", input.Item1);
                return true;
            }
        }

        [Test]
        public async Task ClientRequest_WithUserHeaderMaxAgeAndPrivate_AddsToUserCache()
        {
            // arrange
            var state = new TestState();
            var expectedResponse = state.AddHttpRequest(1);
            expectedResponse.Headers.CacheControl = new CacheControlHeaderValue
            {
                MaxAge = TimeSpan.FromDays(1),
                Private = true
            };

            // act
            var response = await state.ExecuteRequest(user: "my user");

            // assert
            Predicate<Tuple<string, HttpResponseMessage, DateTime>> assert = AssertResult;
            state.Dependencies
                .Verify(x => x.Cache.Put(Match.Create(assert)), Times.Once);

            bool AssertResult(Tuple<string, HttpResponseMessage, DateTime> input)
            {
                Assert.AreEqual("$:my user$:http://www.com/", input.Item1);
                return true;
            }
        }

        [Test]
        public async Task ClientRequest_UsernameHasDollar_EscapesDollarInKey()
        {
            // arrange
            var state = new TestState();
            var expectedResponse = state.AddHttpRequest(1);
            expectedResponse.Headers.CacheControl = new CacheControlHeaderValue
            {
                MaxAge = TimeSpan.FromDays(1),
                Private = true
            };

            // act
            var response = await state.ExecuteRequest(user: "my$user");

            // assert
            Predicate<Tuple<string, HttpResponseMessage, DateTime>> assert = AssertResult;
            state.Dependencies
                .Verify(x => x.Cache.Put(Match.Create(assert)), Times.Once);

            bool AssertResult(Tuple<string, HttpResponseMessage, DateTime> input)
            {
                Assert.AreEqual("$:my$$user$:http://www.com/", input.Item1);
                return true;
            }
        }
        
        [Test]
        public async Task ClientRequest_WithNoHeadersOrCache_ChecksUserAndSharedCache()
        {
            // arrange
            var state = new TestState();
            state.AddHttpRequest(1);

            // act
            var response = await state.ExecuteRequest(user: "my user");

            // assert
            Predicate<string> assertShared = AssertShared;
            Predicate<string> assertUser = AssertUser;
            state.Dependencies
                .Verify(x => x.Cache.Get(It.IsAny<string>()), Times.Exactly(2));
            state.Dependencies
                .Verify(x => x.Cache.Get(Match.Create(assertShared)), Times.Once);
            state.Dependencies
                .Verify(x => x.Cache.Get(Match.Create(assertUser)), Times.Once);

            bool AssertShared(string input)
            {
                return "$:$:http://www.com/" == input;
            }

            bool AssertUser(string input)
            {
                return "$:my user$:http://www.com/" == input;
            }
        }

        [Test]
        public async Task ClientRequest_CachedValueForUserAndShared_ReturnsUserValue()
        {
            // arrange
            var state = new TestState();
            var sharedResponse = state.AddToCache(DateTime.UtcNow.AddDays(1), addResponseContent: 1);
            var userResponse = state.AddToCache(DateTime.UtcNow.AddDays(1), addResponseContent: 2, user: "my user");
            var serverResponse = state.AddHttpRequest(3);

            // act
            var response = await state.ExecuteRequest(user: "my user");

            // assert
            Assert.AreEqual(userResponse, response);
        }
    }
}

File: ShinyHttpCache.Tests/UserCacheTests.fs
module ShinyHttpCache.Tests.UserCacheTests

open System.Net.Http.Headers
open ShinyHttpCache.Model
open ShinyHttpCache.Model.CacheSettings
open ShinyHttpCache.Tests.Utils.AssertUtils
open System
open NUnit.Framework
open ShinyHttpCache.Tests.Utils
open ShinyHttpCache.Tests.Utils.TestState
open ShinyHttpCache.Tests.Utils.TestUtils

[<Test>]
let ``Client request; with user header, max-age, public; Adds to shared cache`` () =
        
    // arrange
    let (_, state, response) =
        TestState.build()
        |> TestState.mockHttpRequest TestState.HttpRequestMock.value
        
    response.Headers.CacheControl <- CacheControlHeaderValue()
    response.Headers.CacheControl.MaxAge <- TimeSpan.FromDays(1.0) |> asNullable
    response.Headers.CacheControl.Private <- false

    // act
    let (mocks, response) =
        TestUtils.ExecuteRequestArgs.value
        |> TestUtils.ExecuteRequestArgs.setUser "my user"
        |> TestUtils.executeRequest
        <| state

    // assert
    async {
        let! _ = response
        
        Mock.Mock.verifyPut (fun (key, _, _) ->
            Assert.AreEqual("G$:$:http://www.com/", key);
            true) mocks
        |> assertEqual 1
        
        return ()
    } |> TestUtils.asTask
    
[<Test>]
let ``Client request; with user header, max-age, private; Adds to user cache`` () =
        
    // arrange
    let (_, state, response) =
        TestState.build()
        |> TestState.mockHttpRequest TestState.HttpRequestMock.value
        
    response.Headers.CacheControl <- CacheControlHeaderValue()
    response.Headers.CacheControl.MaxAge <- TimeSpan.FromDays(1.0) |> asNullable
    response.Headers.CacheControl.Private <- true

    // act
    let (mocks, response) =
        TestUtils.ExecuteRequestArgs.value
        |> TestUtils.ExecuteRequestArgs.setUser "my user"
        |> TestUtils.executeRequest
        <| state

    // assert
    async {
        let! _ = response
        
        Mock.Mock.verifyPut (fun (key, _, _) ->
            Assert.AreEqual("G$:my user$:http://www.com/", key);
            true) mocks
        |> assertEqual 1
        
        return ()
    } |> TestUtils.asTask
    
[<Test>]
let ``Client request; user name has $; Escapes dollar`` () =
        
    // arrange
    let (_, state, response) =
        TestState.build()
        |> TestState.mockHttpRequest TestState.HttpRequestMock.value
        
    response.Headers.CacheControl <- CacheControlHeaderValue()
    response.Headers.CacheControl.MaxAge <- TimeSpan.FromDays(1.0) |> asNullable
    response.Headers.CacheControl.Private <- true

    // act
    let (mocks, response) =
        TestUtils.ExecuteRequestArgs.value
        |> TestUtils.ExecuteRequestArgs.setUser "my$user"
        |> TestUtils.executeRequest
        <| state

    // assert
    async {
        let! _ = response
        
        Mock.Mock.verifyPut (fun (key, _, _) ->
            Assert.AreEqual("G$:my$$user$:http://www.com/", key);
            true) mocks
        |> assertEqual 1
        
        return ()
    } |> TestUtils.asTask
    
[<Test>]
let ``Client request; No headers or cache; Checks user and shared cache`` () =
        
    // arrange
    let (_, state, _) =
        TestState.build()
        |> TestState.mockHttpRequest TestState.HttpRequestMock.value

    // act
    let (mocks, response) =
        TestUtils.ExecuteRequestArgs.value
        |> TestUtils.ExecuteRequestArgs.setUser "my user"
        |> TestUtils.executeRequest
        <| state

    // assert
    async {
        let! _ = response
        
        Mock.Mock.verifyGet (fun _ -> true) mocks
        |> assertEqual 2
        Mock.Mock.verifyGet ((=) "G$:my user$:http://www.com/") mocks
        |> assertEqual 1
        Mock.Mock.verifyGet ((=) "G$:$:http://www.com/") mocks
        |> assertEqual 1
        
        return ()
    } |> TestUtils.asTask
    
[<Test>]
let ``Client request; Cached value for user and shared; Returns user value`` () =
        
    // arrange
    let cached1 =
        TestState.CachedData.value (DateTime.UtcNow.AddDays(1.0) |> CachedData.Expires)
        |> TestState.CachedData.setResponseContent 99
    let cached2 =
        TestState.CachedData.value (DateTime.UtcNow.AddDays(1.0) |> CachedData.Expires)
        |> TestState.CachedData.setResponseContent 77
        |> TestState.CachedData.setUser "my user"
        
    let (_, state, _) =
        TestState.build()
        |> TestState.mockHttpRequest TestState.HttpRequestMock.value
    let (_, state) = TestState.addToCache cached1 state
    let (_, state) = TestState.addToCache cached2 state

    // act
    let (mocks, response) =
        TestUtils.ExecuteRequestArgs.value
        |> TestUtils.ExecuteRequestArgs.setUser "my user"
        |> TestUtils.executeRequest
        <| state

    // assert
    async {
        let! response = response
        do! assertContent 77 response
        
        return ()
    } |> TestUtils.asTask